enhancements/installer/aws-custom-dns.md

```diff
@@ -0,0 +1,544 @@
+api-int and *.apps services, then they could optionally delete the in-cluster
+CoreDNS pod and the cluster is expected to function fully as expected. This is
+a completely optional step with this design. If the customer does configure
+their custom DNS solution and leave the in-cluster CoreDNS pod running, all
+in-cluster entities will end up using the CoreDNS pod's DNS services and all
+out of cluster requests will be handled by the external DNS.
+
+### Workflow Description
+
+1. For an OpenShift administrator to use a custom DNS solution, they would
+have to enable a config option in the install configuration. This would
+cause the Installer and the Ingress Operator to skip configuring the cloud
+DNS with entries for API, Internal API and Ingress Services. 
+
+2. The Installer and the Ingress Operator continue to configure the cloud LBs
+on the public subnets of Public Hosted Zones for `api` and `*.apps`. They will
+continue to configure cloud LBs on the private subnets of Private Hosted Zones
+for `api-int` and `*.apps` for in-cluster access to the Ingress service.
+
+3. The Installer will also update the new fields `APIServerDNSConfig` and
+`InternalAPIServerDNSConfig` within the `AWSPlatformStatus`, with the LB
+IP addresses and the DNS Record types for API, API-Int servers respectively.
+
+4. The Installer instructs the machine-config-operator(MCO) to start a static
+CoreDNS pod on the bootstrap and master nodes to provide DNS resolution for
+the API and API-Int servers. MCO uses the DNS config in the `AWSPlatformStatus`
+to generate the CoreFile to configure the CoreDNS instance in the pod.
+Note: This implementation differs from the on-prem case by not staring the
+keepalived or haproxy containers for load balancing.
+
+5. The Installer omits the DNS Zones within `dnses.config.openshift.io/cluster`
+to influence the behavior of the Ingress Operator by making it skip the
+configuration of cloud DNS while continuing with LB configuration.
+
+6. The Ingress Controllers managed by the cluster-ingress-operator continue
+to configure the application LBs to enable in-cluster and external accesss to
+cluster services. The Ingress Controllers will continue to update the DNSRecord
+ConfigResource(CR) with the configuration required to configure DNS for these
+services.
+
+7. The MCO will watch the DNSRecord CR for the DNS configuration mentioned in
+step 6. MCO will update the newly added `IngressDNSConfig` field within the
+`AWSPlatformStatus` with the DNS configuartion for *.apps records by reading
+the DNSRecord CR.
+
+8. The runtimecfg utility will monitor the `Infrastructure` CR for changes to
+`IngressDNSConfig` of `AWSPlatformStatus` to update the CoreDNS pod's
+CoreFile to now also include DNS configuration for both internal and
+external access to the cluster services (the `*.apps` records).
+
+9. Once the cluster is functional, the user can look at the contents of the new
```

Comment:
I think prior to this step, we should add that the installer would need to be connecting to the API using the LB DNS name, since the DNS records for the api have not been established yet.

I think we have an open question about whether this is technically feasible, specifically with the generated certs.

Reply:
There is no need to do that. The CoreDNS pod would be providing API server resolution during cluster installation.

Reply:
There is no issue with the generated certs. Inter-pod communication happens with certs generated for API-Int which will always be resolvable by the in-cluster DNS.
The certs generated for api and *.apps are generated using "api.<cluster domain name> / *.apps.<cluster domain name>. That remains the same no-matter which DNS ends up resolving the request, the in-cluster one or the external one.

Reply:
Ack. This might become an exercise in figuring out what the current installer code will allow and less about what is the best solution for the customer. Will look through diff options on how this issue can be averted and will report back.

Reply:
>So I presume we would need to change the installer code so that the installer queries the LB DNS name directly. That is also why I was wondering about the certificates. I hope that makes sense!

You can pass in a dial configuration when you create a kubernetes client to dial to a different IP address than listed in the server.  An example into what was implemented in hive is here: https://github.com/openshift/hive/pull/1817/files

Should be possible to extend this to use the DNS label on the load balancer or an IP address.